pkg/client/kubernetes/cache/single_object_test.go

```diff
@@ -0,0 +1,183 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache_test
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	testclock "k8s.io/utils/clock/testing"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	. "github.com/gardener/gardener/pkg/client/kubernetes/cache"
+	mockcache "github.com/gardener/gardener/pkg/mock/controller-runtime/cache"
+)
+
+var _ = Describe("SingleObject", func() {
+	var (
+		ctx  context.Context
+		ctrl *gomock.Controller
+
+		obj    *corev1.Secret
+		key    client.ObjectKey
+		resync = pointer.Duration(time.Second)
+
+		mockCache         *mockcache.MockCache
+		singleObjectCache cache.Cache
+		clock             *testclock.FakeClock
+		maxIdleTime       time.Duration
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+
+		ctrl = gomock.NewController(GinkgoT())
+		DeferCleanup(func() {
+			ctrl.Finish()
+		})
+
+		obj = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "foo",
+				Namespace: "bar",
+			},
+		}
+		key = client.ObjectKeyFromObject(obj)
+
+		mockCache = mockcache.NewMockCache(ctrl)
+		newCacheFunc := func(_ *rest.Config, opts cache.Options) (cache.Cache, error) {
+			Expect(opts.Namespace).To(Equal(obj.Namespace))
+			Expect(opts.DefaultSelector).To(Equal(cache.ObjectSelector{Field: fields.SelectorFromSet(fields.Set{"metadata.name": obj.Name})}))
+			Expect(opts.SelectorsByObject).To(BeNil())
+			Expect(opts.Resync).To(Equal(resync))
+			return mockCache, nil
+		}
+		clock = testclock.NewFakeClock(time.Now())
+
+		maxIdleTime = 100 * time.Millisecond
+		singleObjectCache = NewSingleObject(logr.Discard(), nil, cache.Options{Resync: resync}, newCacheFunc, clock, maxIdleTime, 50*time.Millisecond)
+
+		var wg wait.Group
+		ctx, cancel := context.WithCancel(ctx)
+
+		DeferCleanup(func() {
+			// cancel ctx to stop garbage collector
+			cancel()
+			// wait for singleObjectCache.Start to return, otherwise test is not finished
+			wg.Wait()
+		})
+
+		wg.Start(func() {
+			defer GinkgoRecover()
+			Expect(singleObjectCache.Start(ctx)).To(Succeed())
+		})
+	})
+
+	Describe("#Get", func() {
+		It("should successfully delegate to stored cache", func() {
+			By("cache does not exist yet")
+			var (
+				cacheCtx  context.Context
+				startChan = make(chan struct{})
+			)
+			mockCache.EXPECT().Start(gomock.Any()).DoAndReturn(func(ctx context.Context) error {
+				cacheCtx = ctx
+				startChan <- struct{}{}
+				return nil
+			})
```

Comment:
Can you use the `expectStart` function here as well?

Reply:
We need to save the `cacheCtx` which is not possible with `expectStart`.

Reply:
For this test, `cacheCtx` is always equal to `ctx`. We don't really need to store it in the `DoAndReturn` function but could simply set `cacheCtx = ctx` in the begining. Wouldn't this work as well?

Reply:
I don't understand. The `SingleObject` cache creates a [new sub context](https://github.com/gardener/gardener/blob/4a40c9fbad0a9efffe31522f704c9038db0c2026/pkg/client/kubernetes/cache/single_object.go#L146) that we need to save in order to check if it has been [closed](https://github.com/gardener/gardener/blob/4a40c9fbad0a9efffe31522f704c9038db0c2026/pkg/client/kubernetes/cache/single_object_test.go#L130). Am I missing something?

Reply:
Right, I was confused by the "duplicative" variable naming `ctx` in the function part of `DoAndReturn` :) Anyways, all good, please keep it this way. Thanks!